Add libzip recipe

ceholden opened the pull request:
Add recipe for [libzip](https://libzip.org/), a C library for reading/writing/modifying zip archives. This recipe is being added primarily as a dependency for upgrading to QGIS 3 (see https://github.com/conda-forge/qgis-feedstock/issues/17).

conda-forge-linter commented:
Hi! This is the friendly automated conda-forge-linting service.

I just wanted to let you know that I linted all conda-recipes in your PR (```recipes/libzip```) and found it was in an excellent condition.

ceholden commented:
The build includes a number of test programs, but some of these have been disabled via the submitted patches because of issues on compiling on Windows. When possible, I tried to disable these tests from being compiled using patches based on commits that the library authors have published since the last stable release. Others who have packed this library for Windows went as far as to completely disable building these test/example programs (e.g., from [Microsoft/vcpkg](https://github.com/Microsoft/vcpkg/blob/master/ports/libzip/cmake_dont_build_more_than_needed.patch)). The test/example programs that haven't been patched out compiled successfully on a local Windows VM, and it seems worth it to build _some_ of these programs as tests despite the added complexity (versus disabling everything).

I'm not very familiar on Windows and welcome any comments/suggestions.

ceholden commented:
The x64 Appveyor build seems to be getting stuck on extracting the source archive, several times in a row now. Maybe it's caching a corrupted archive and uses that instead of downloading anew? The x86 build passed so hopefully this is just a minor annoyance.

```
Source cache directory is: C:\bld\src_cache
Downloading source to cache: libzip-1.4.0.tar.gz
Downloading https://libzip.org/download/libzip-1.4.0.tar.xz
libzip-1.4.0.t 100% |###############################| Time: 0:00:01 552.99 kB/s
Success
Traceback (most recent call last):
  File "C:\Miniconda-x64\Scripts\conda-build-all-script.py", line 5, in <module>
    exit(main())
  File "C:\Miniconda-x64\lib\site-packages\conda_build_all\cli.py", line 93, in main
    b.main()
  File "C:\Miniconda-x64\lib\site-packages\conda_build_all\builder.py", line 297, in main
    built_dist_location = self.build(meta, build_config)
  File "C:\Miniconda-x64\lib\site-packages\conda_build_all\builder.py", line 227, in build
    output_paths = conda_build.api.build(meta.meta, config=config)
  File "C:\Miniconda-x64\lib\site-packages\conda_build\api.py", line 97, in build
    need_source_download=need_source_download, config=config)
  File "C:\Miniconda-x64\lib\site-packages\conda_build\build.py", line 1524, in build_tree
    config=config)
  File "C:\Miniconda-x64\lib\site-packages\conda_build\build.py", line 1062, in build
    config=config)
  File "C:\Miniconda-x64\lib\site-packages\conda_build\render.py", line 86, in parse_or_try_download
    source.provide(metadata, config=config)
  File "C:\Miniconda-x64\lib\site-packages\conda_build\source.py", line 519, in provide
    unpack(metadata, config=config)
  File "C:\Miniconda-x64\lib\site-packages\conda_build\source.py", line 91, in unpack
    tar_xf(src_path, config.work_dir)
  File "C:\Miniconda-x64\lib\site-packages\conda_build\utils.py", line 340, in tar_xf
    t = tarfile.open(tarball, mode)
  File "C:\Miniconda-x64\lib\tarfile.py", line 1680, in open
    raise ReadError("file could not be opened successfully")
tarfile.ReadError: file could not be opened successfully
Command exited with code 1
```
Any suggestions?

isuruf commented on recipes/libzip/meta.yaml:
```diff
@@ -0,0 +1,75 @@
+  version: {{ version }}
+
+source:
+  fn: {{ name }}-{{ version }}.tar.gz
```
`.tar.xz`

ceholden commented:
I rebased to include recent commits, and squashed my work in progress commits down to 1 commit. Assuming it's still building correctly, I would really appreciate a review and merge so I can get to updating the QGIS recipe that depends on this library. Thank you!

dougalsutherland commented on recipes/libzip/meta.yaml:
```diff
@@ -0,0 +1,75 @@
+    - vc 14  # [win and py35]
+    - vc 14  # [win and py36]
+    # perl is needed to run the package tests
+    - perl
```
This is going to cause some problems while we're still on conda-build 2, because it does some automatic pinning stuff for perl. If you make it `perl >0`, it'll work correctly (https://github.com/conda-forge/parallel-feedstock/pull/1#issuecomment-362510209).

dougalsutherland commented on recipes/libzip/meta.yaml:
```diff
@@ -0,0 +1,75 @@
+requirements:
+  build:
+    - cmake
+    - zlib
```
Should be `zlib 1.2.11` while we're still on conda-build 2.

dougalsutherland commented on recipes/libzip/meta.yaml:
```diff
@@ -0,0 +1,75 @@
+  build:
+    - cmake
+    - zlib
+    - bzip2
```
Same here, `bzip2 1.0`.

dougalsutherland commented on recipes/libzip/meta.yaml:
```diff
@@ -0,0 +1,75 @@
+    # perl is needed to run the package tests
+    - perl
+  run:
+    - zlib
```
`zlib 1.2.11`

dougalsutherland commented on recipes/libzip/meta.yaml:
```diff
@@ -0,0 +1,75 @@
+    - perl
+  run:
+    - zlib
+    - bzip2
```
`bzip2 1.0`

ceholden commented:
@dougalsutherland thank you for the review and comments. The updated recipe passed all CIs if you want to make the changes and merge.

I forgot about those pins, and will add them in to the other PR I have going :smiley:

dougalsutherland commented:
Merged! Thanks @ceholden.

For the record, I tested locally and `perl >0` isn't currently enough in this case; need to pin a specific version. :/
